Big Data And Society: Lab 2
=====


# Forking the Repository

If you already have a copy of the repo, you can jump down to the next step.

First, let's fork the course repo. A fork is a copy of a repository. Unlike cloning, when you fork a repo, it is attached to your account, and you are free to modify it, and edit, without affecting the original version... which is desirable for the context of the class.

1. First click fork on the github page
2. On GitHub, navigate to your fork of repository.
3. In the right sidebar of your fork's repository page, click  to copy the clone URL for your fork.
4. Open Terminal (for Mac and Linux users) or the command prompt (for Windows users).
5. Type git clone, and then paste the URL you copied in Step 2. 

### Updating your Version to get Materials from the Main Class Repository

To get the new materials from the main class repository, you need to commit any changes you've made locally and push them to your personal fork of the repo online. Then pull the repository from the main class repo.

1. Type git remote add upstream, and then paste the URL https://github.com/civic-data-design-lab/big-data-spring2017.git (you might get a message saying that this already exists, this is fine, it means this has already been done.)
2. To verify the new upstream repository you've specified for your fork, type **git remote -v**. You should see the URL for your fork as origin (remember your fork has **your username** in it), and the URL for the original repository as upstream (the one with the **civic-data-design-lab** in the URL).
3. Before you pull the upstream master repo, commit any changes you've made locally by following these steps:

===

* Stage commits using **git add .**
* Commit your staged changes using **git commit -m "my changes"**
* Push your work to your forked repo. **git push origin master** (if you get a warning about LF conflicting with CRLF, this is not a problem, you can proceed)

===
    
This will update your forked repo with any local changes.
    
4. Once this is done, **pull** the upstream repository (the main class repository). This will both fetch and merge the original repo into your local version. Do this using **git pull upstream master**

Pulling from the remote repository will bring in its branches and their respective commits.

# Introduction to Pandas
![IPython](images/pandas.png "Pandas")
**Pandas** is a library for Python for data manipulation and analysis. Pandas expands the data processing capacities of Python and adds a number of classes for easily importing data, in particular numerical tables, from various formats into their DataFrame object. A `DataFrame` is Panda’s basic object that allows multidimensional data processing and indexing. `DataFrames` can be easily and efficiently queried without the need of cumbersome syntax and convoluted loops. `DataFrames` can be merged with other data, they can be sliced, and they can be reshaped; in a way, we can think of Pandas as a big data combination of Excel and SQL. 

## Resources

Pandas employs a number of functional and declarative programming idioms making it a bit different from regulay Python syntax. However, it is very close to functionality to **NumPy**, treating objects as vectors. In this exercise, **NumPy** is generally used for math functions.

You can get a Cheatsheet to ease the learning curve, and learn about some of its functionality:

https://drive.google.com/folderview?id=0ByIrJAE4KMTtaGhRcXkxNHhmY2M&usp=sharing




### 0. Importing the library
First we need to import the library, and use a shorter name for it:

In [1]:
import pandas as pd
import numpy as np

### 1. Pandas and Data Frames

You can think of a dataframe as a manipulable, multidimension Python array. They are based off of numpy arrays, can hold structured data, and you can perform functions on them to manage data, such as query, select, join, group. Let's setup a really simple dataframe.

In [2]:
# Create empty dataframe
df = pd.DataFrame()

# Create a column
df['name'] = ['Joey', 'Jeremy', 'Jenny']

# View dataframe
df

name
0    Joey
1  Jeremy
2   Jenny

Say we want to add a column, its easy.

In [3]:
# Assign a new column to df called 'age' with a list of ages
df.assign(age = [28, 33, 27])

name  age
0    Joey   28
1  Jeremy   33
2   Jenny   27

Data in this table can be manipulated, queried, sorted, and munged. Removing this for an inclass demo!

### 2. Reading Files

This is cool, but you don't normally load data like this though, it gets tedious!

Pandas provides a number of reader functions that process files and return a pandas object or `DataFrame`. Multiple different file types can be read, like **csv**, **txt**, **xls**, and **json**. The function parses the tabular data contained in the files and returns a formatted and indexed `DataFrame`. A number of additional arguments can be speficied, allowing to define the type of separator for text files, the header, the column to use as an index, etc. 

Additional documentation can be found on the API: 
http://pandas.pydata.org/pandas-docs/stable/index.html

To start us off, lets look at a results file by precinct from the 2016 general election. This is an extract of results by precinct for Centre County, Pennsylvania.


In [4]:
# We are reading a CSV with election data
df = pd.read_table('data/centrecounty_precinct_results_nov2016.txt', sep=',')

# We can print the first 5 rows of the df
df.head()

County  ElecYear          Election         ElecDate  \
0  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
1  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
2  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
3  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
4  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   

   PrecNo             PrecName       Party                    Contest  Descr  \
0     1.0  01 BELLEFONTE NORTH         NaN  REGISTERED VOTERS - TOTAL    NaN   
1     1.0  01 BELLEFONTE NORTH         NaN       BALLOTS CAST - TOTAL    NaN   
2     1.0  01 BELLEFONTE NORTH         NaN      VOTER TURNOUT - TOTAL    NaN   
3     1.0  01 BELLEFONTE NORTH  DEMOCRATIC             STRAIGHT PARTY    NaN   
4     1.0  01 BELLEFONTE NORTH  REPUBLICAN             STRAIGHT PARTY    NaN   

                     VoteFor   Candidate  Posn  Count  PctCnt  Total  PctTot  
0                        NaN         NaN  10.0  507.0     NaN    NaN     NaN  
1                        NaN         NaN  11.0  391.0     NaN    NaN     NaN  
2                        NaN         NaN  12.0    NaN   77.12    NaN     NaN  
3  VOTE FOR NOT MORE THAN  1  DEMOCRATIC  17.0   46.0   40.35    NaN     NaN  
4  VOTE FOR NOT MORE THAN  1  REPUBLICAN  18.0   66.0   57.89    NaN     NaN

Notice we have a table! A spreadsheet! And it indexed the rows. Pandas (borrowing from R) calls it a DataFrame. Lets see the types of the columns.

`df`, in Python context, is an instance of the `pd.DataFrame` class, created by calling the `pd.read_table` function, which calls the DataFrame constructor inside of it. `df` is a dataframe object.



### 2. DataFrame methods:
The df object, for example, has methods, or functions belonging to it, which allow it to do things. For example `df.head()` is a method that shows the first 5 rows of the dataframe. `df.dtypes` returns the data types of each of the columns. 

Additional documentation can be found here: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html
In general, the API is the best resource: http://pandas.pydata.org/pandas-docs/stable/api.html

In [5]:
df.dtypes

County        object
ElecYear     float64
Election      object
ElecDate      object
PrecNo       float64
PrecName      object
Party         object
Contest       object
Descr        float64
VoteFor       object
Candidate     object
Posn         float64
Count        float64
PctCnt       float64
Total        float64
PctTot       float64
dtype: object

The shape of the object is:

In [6]:
df.shape

(3526, 16)

3526 rows times 16 columns. A spredsheet is a matrix. `df.shape` returns a tuple, so we can access members of this **tuple** like we do with a list:

In [7]:
df.shape[0], df.shape[1]

(3526, 16)

In [8]:
df.shape[0]

3526

These are the column names:

In [9]:
df.columns

Index(['County', 'ElecYear', 'Election', 'ElecDate', 'PrecNo', 'PrecName',
       'Party', 'Contest', 'Descr', 'VoteFor', 'Candidate', 'Posn', 'Count',
       'PctCnt', 'Total', 'PctTot'],
      dtype='object')

Notice that `df.columns` returns a `Pandas.Series` object. This object is built on top of the Python **lists**, and has similar methods and attributes. 

Access an individual column.

In [10]:
df.Contest

0                               REGISTERED VOTERS - TOTAL
1                                    BALLOTS CAST - TOTAL
2                                   VOTER TURNOUT - TOTAL
3                                          STRAIGHT PARTY
4                                          STRAIGHT PARTY
5                                          STRAIGHT PARTY
6                                          STRAIGHT PARTY
7                                          STRAIGHT PARTY
8                                   PRESIDENTIAL ELECTORS
9                                   PRESIDENTIAL ELECTORS
10                                  PRESIDENTIAL ELECTORS
11                                  PRESIDENTIAL ELECTORS
12                                  PRESIDENTIAL ELECTORS
13                                  PRESIDENTIAL ELECTORS
14                                  UNITED STATES SENATOR
15                                  UNITED STATES SENATOR
16                                  UNITED STATES SENATOR
17            

We can use the unique() function to find the unique values in each column. Let's find the unique values in the Contest column

In [11]:
df.Contest.unique()

array(['REGISTERED VOTERS - TOTAL', 'BALLOTS CAST - TOTAL',
       'VOTER TURNOUT - TOTAL', 'STRAIGHT PARTY', 'PRESIDENTIAL ELECTORS',
       'UNITED STATES SENATOR', 'ATTORNEY GENERAL', 'AUDITOR GENERAL',
       'STATE TREASURER', 'REPRESENTATIVE IN CONGRESS 5TH CONGRESSIONAL',
       'REPRESENTATIVE IN THE GENERAL ASSEMBLY 171ST LEGISLATIVE',
       'JUDGES AGE',
       'REPRESENTATIVE IN THE GENERAL ASSEMBLY 76TH LEGISLATIVE',
       'REPRESENTATIVE IN THE GENERAL ASSEMBLY 77TH LEGISLATIVE',
       'REPRESENTATIVE IN THE GENERAL ASSEMBLY 81ST LEGISLATIVE', nan], dtype=object)

Now lets get multiple columns:

In [12]:
df_multipleColumns = df[['Contest','PrecNo', 'Count']]
df_multipleColumns.head()

Contest  PrecNo  Count
0  REGISTERED VOTERS - TOTAL     1.0  507.0
1       BALLOTS CAST - TOTAL     1.0  391.0
2      VOTER TURNOUT - TOTAL     1.0    NaN
3             STRAIGHT PARTY     1.0   46.0
4             STRAIGHT PARTY     1.0   66.0

### 3. Querying

Pandas `DataFrames` have built-in methods for performing queries in a SQL style. They can be diced, sorted, etc it. We can apply any of this queries to parts of the `df`, and based on this query select subsets of the whole `df`. Let's do some filtering and querying.

The output of this is a series.

In [13]:
df['PrecNo'] == 1

0        True
1        True
2        True
3        True
4        True
5        True
6        True
7        True
8        True
9        True
10       True
11       True
12       True
13       True
14       True
15       True
16       True
17       True
18       True
19       True
20       True
21       True
22       True
23       True
24       True
25       True
26       True
27       True
28       True
29       True
        ...  
3496    False
3497    False
3498    False
3499    False
3500    False
3501    False
3502    False
3503    False
3504    False
3505    False
3506    False
3507    False
3508    False
3509    False
3510    False
3511    False
3512    False
3513    False
3514    False
3515    False
3516    False
3517    False
3518    False
3519    False
3520    False
3521    False
3522    False
3523    False
3524    False
3525    False
Name: PrecNo, dtype: bool

To print out the table with only the rows we want, we can apply what is called a filter. The above is a series. Let's take a look again at **Precinct 1**. We can use the `df` mask to get a filtered dataframe: we use the mask to "index" into the dataframe to get the rows we want.

In [14]:
df_precinct1 = df[df['PrecNo'] == 1]

In [15]:
df_precinct1.head()

County  ElecYear          Election         ElecDate  \
0  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
1  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
2  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
3  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
4  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   

   PrecNo             PrecName       Party                    Contest  Descr  \
0     1.0  01 BELLEFONTE NORTH         NaN  REGISTERED VOTERS - TOTAL    NaN   
1     1.0  01 BELLEFONTE NORTH         NaN       BALLOTS CAST - TOTAL    NaN   
2     1.0  01 BELLEFONTE NORTH         NaN      VOTER TURNOUT - TOTAL    NaN   
3     1.0  01 BELLEFONTE NORTH  DEMOCRATIC             STRAIGHT PARTY    NaN   
4     1.0  01 BELLEFONTE NORTH  REPUBLICAN             STRAIGHT PARTY    NaN   

                     VoteFor   Candidate  Posn  Count  PctCnt  Total  PctTot  
0                        NaN         NaN  10.0  507.0     NaN    NaN     NaN  
1                        NaN         NaN  11.0  391.0     NaN    NaN     NaN  
2                        NaN         NaN  12.0    NaN   77.12    NaN     NaN  
3  VOTE FOR NOT MORE THAN  1  DEMOCRATIC  17.0   46.0   40.35    NaN     NaN  
4  VOTE FOR NOT MORE THAN  1  REPUBLICAN  18.0   66.0   57.89    NaN     NaN

We can query based on multiple conditions, by using a `&` condition. All we need to do is add `()` brackets around each condition. The query uses a boolean AND. Each condition ceates a mask of trues and falses.

In [16]:
df[(df['PrecNo'] < 8) & (df['PrecNo'] > 5)]

County  ElecYear          Election         ElecDate  \
195  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
196  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
197  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
198  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
199  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
200  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
201  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
202  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
203  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
204  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
205  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
206  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
207  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
208  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
209  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
210  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
211  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
212  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
213  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
214  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
215  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
216  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
217  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
218  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
219  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
220  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
221  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
222  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
223  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
224  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
..                           ...       ...               ...              ...   
243  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
244  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
245  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
246  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
247  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
248  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
249  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
250  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
251  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
252  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
253  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
254  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
255  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
256  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
257  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
258  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
259  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
260  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECT

Because this dataset contains all of the different contests in the November generals, we can use a query to pull out just the presidential election. Here we use [df.loc](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html) to select by label and find only those listings for the presidential election.

In [17]:
df_pres = df.loc[df['Contest'] == 'PRESIDENTIAL ELECTORS']
df_pres

County  ElecYear          Election  \
8     CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
9     CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
10    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
11    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
12    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
13    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
47    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
48    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
49    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
50    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
51    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
52    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
86    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
87    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
88    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
89    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
90    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
91    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
125   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
126   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
127   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
128   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
129   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
130   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
164   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
165   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
166   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
167   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
168   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
169   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
...                           ...       ...               ...   
3333  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3334  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3335  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3336  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3337  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3338  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3372  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3373  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3374  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3375  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3376  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3377  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3411  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3412  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3413  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3414  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3415  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3416  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3450  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3451  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3452  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3453  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3454  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3455  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3489  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3490  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3491  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3492  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3493  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   
3494  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION 

Now that we are looking just at the votes for the presidential election, we can find, for example, which rows show votes for Hillary Clinton.

In [18]:
df_clinton = df_pres.loc[df_pres['Candidate'] == 'HILLARY CLINTON,  PRESIDENT']
df_clinton.head()

County  ElecYear          Election         ElecDate  \
8    CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
47   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
86   CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
125  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   
164  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   

     PrecNo                 PrecName       Party                Contest  \
8       1.0      01 BELLEFONTE NORTH  DEMOCRATIC  PRESIDENTIAL ELECTORS   
47      2.0  02 BELLEFONTE NORTHEAST  DEMOCRATIC  PRESIDENTIAL ELECTORS   
86      3.0      03 BELLEFONTE SOUTH  DEMOCRATIC  PRESIDENTIAL ELECTORS   
125     4.0  04 BELLEFONTE SOUTHEAST  DEMOCRATIC  PRESIDENTIAL ELECTORS   
164     5.0       05 BELLEFONTE WEST  DEMOCRATIC  PRESIDENTIAL ELECTORS   

     Descr                    VoteFor                    Candidate   Posn  \
8      NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT   26.0   
47     NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  111.0   
86     NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  196.0   
125    NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  281.0   
164    NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  366.0   

     Count  PctCnt  Total  PctTot  
8    188.0   48.08    NaN     NaN  
47   259.0   38.14    NaN     NaN  
86   230.0   50.33    NaN     NaN  
125  283.0   39.47    NaN     NaN  
164  443.0   48.05    NaN     NaN

In [19]:
df_pres['Candidate'].unique()

array(['HILLARY CLINTON,  PRESIDENT', 'DONALD J TRUMP,  PRESIDENT',
       'DARRELL L CASTLE,  PRESIDENT', 'JILL STEIN,  PRESIDENT',
       'GARY JOHNSON, PRESIDENT', 'WRITE-IN'], dtype=object)

Since these rows are the ones that have voted for Clinton, we can perform operations to locate which rows (ie which precincts) had more than 50% of the vote go to Clinton.

This gives us a `Pandas.Series` of `True`s and `False`s. We call this a mask.

In [20]:
df_clinton.PctCnt > 50

8       False
47      False
86       True
125     False
164     False
203     False
242     False
281     False
320     False
359     False
397     False
435     False
473     False
512     False
551     False
589      True
627      True
665      True
703      True
741      True
779      True
817      True
855      True
893      True
931      True
969      True
1007     True
1045     True
1083     True
1121     True
        ...  
2364    False
2403    False
2442     True
2480    False
2519     True
2558     True
2596     True
2635    False
2674    False
2713    False
2752    False
2790    False
2828    False
2866    False
2904    False
2943    False
2982    False
3021    False
3060    False
3099    False
3138    False
3177    False
3216    False
3255    False
3294    False
3333    False
3372     True
3411     True
3450    False
3489     True
Name: PctCnt, dtype: bool

We can summarize the Count field for the rows that represent Clinton.

In [21]:
df_clinton.Count.sum()

37088.0

Say we want to summary for all of the candidates, lets go back to the **df_pres** table we made and group by the candidate, then sum the count.

In [22]:
df_pres.groupby('Candidate').Count.sum()

Candidate
DARRELL L CASTLE,  PRESIDENT      344.0
DONALD J TRUMP,  PRESIDENT      35274.0
GARY JOHNSON, PRESIDENT          2644.0
HILLARY CLINTON,  PRESIDENT     37088.0
JILL STEIN,  PRESIDENT            798.0
WRITE-IN                         1159.0
Name: Count, dtype: float64

We can group by with multiple columns. This will allow us to look at the results from each precinct.

In [23]:
df_pres.groupby(['PrecNo','Candidate']).Count.sum()

PrecNo  Candidate                   
1.0     DARRELL L CASTLE,  PRESIDENT      2.0
        DONALD J TRUMP,  PRESIDENT      174.0
        GARY JOHNSON, PRESIDENT          22.0
        HILLARY CLINTON,  PRESIDENT     188.0
        JILL STEIN,  PRESIDENT            3.0
        WRITE-IN                          2.0
2.0     DARRELL L CASTLE,  PRESIDENT      9.0
        DONALD J TRUMP,  PRESIDENT      385.0
        GARY JOHNSON, PRESIDENT          10.0
        HILLARY CLINTON,  PRESIDENT     259.0
        JILL STEIN,  PRESIDENT            7.0
        WRITE-IN                          9.0
3.0     DARRELL L CASTLE,  PRESIDENT      6.0
        DONALD J TRUMP,  PRESIDENT      193.0
        GARY JOHNSON, PRESIDENT          17.0
        HILLARY CLINTON,  PRESIDENT     230.0
        JILL STEIN,  PRESIDENT            4.0
        WRITE-IN                          7.0
4.0     DARRELL L CASTLE,  PRESIDENT      6.0
        DONALD J TRUMP,  PRESIDENT      375.0
        GARY JOHNSON, PRESIDENT          35

Or directly, in Pandas, which works since `df_pres` is a pandas Series. Pandas series have a number of built-in methods like `.max()`, `.min()`, and `.mean()`!

In [24]:
df_clinton['Count'].max()

1195.0

You can save these as variables. And cast them to strings.

In [25]:
# create min, max, and mean
max_clinton = df_clinton['Count'].max()
min_clinton = df_clinton['Count'].min()
mean_clinton = df_clinton['Count'].mean()

# print out some results
print("Highest precinct vote total for Clinton: " + str(max_clinton))
print("Lowest precinct vote total for Clinton: " + str(min_clinton))
print("Mean precinct vote total for Clinton: " + str(mean_clinton))

Highest precinct vote total for Clinton: 1195.0
Lowest precinct vote total for Clinton: 18.0
Mean precinct vote total for Clinton: 407.56043956


This isn't super useful though, so we can get the whole row of the max and min values of the percentage vote, use the following query.

In [26]:
df_clinton.loc[df['PctCnt'] == df_clinton['PctCnt'].max()]

County  ElecYear          Election         ElecDate  \
589  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION  NOVEMBER 8 2016   

     PrecNo     PrecName       Party                Contest  Descr  \
589    16.0  16 SC NORTH  DEMOCRATIC  PRESIDENTIAL ELECTORS    NaN   

                       VoteFor                    Candidate    Posn  Count  \
589  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  1297.0  495.0   

     PctCnt  Total  PctTot  
589   77.95    NaN     NaN

Here, Clinton took 77% of the vote!

In [27]:
df_clinton.loc[df['PctCnt'] == df_clinton['PctCnt'].min()]

County  ElecYear          Election  \
3138  CENTRE COUNTY, PENNSYLVANIA    2016.0  GENERAL ELECTION   

             ElecDate  PrecNo  PrecName       Party                Contest  \
3138  NOVEMBER 8 2016    82.0  82 UNION  DEMOCRATIC  PRESIDENTIAL ELECTORS   

      Descr                    VoteFor                    Candidate    Posn  \
3138    NaN  VOTE FOR NOT MORE THAN  1  HILLARY CLINTON,  PRESIDENT  6886.0   

      Count  PctCnt  Total  PctTot  
3138  126.0   17.03    NaN     NaN

In this precinct, Clinton took only 17% of the votes.

#### Exercise:

Calculate the minimum value, and maximum value of the `Count` column to find the precincts with the highest and lowest turnouts by percentage. Hint, you'll want to filter by `Contest`. [This](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.min.html#pandas.Series.min) documentation might be useful.

In [28]:
# Your code here


#### Exercise:
Get the average actual number of people that turned out to vote by precinct across the whole dataset. [Documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.subtract.html#pandas.Series.subtract)

In [29]:
# Your code here


### 5. Cleaning

One of the most common tasks while working with big data is data cleaning. Datasets will be inherently heterogeneous and unstructured. The lack of structure can cause functions to throw errors. We can clean data sets through various methods; if the discrepancies follow a structured pattern, it is possible to use built-in functions, or write our own functions. However, if the errors cannot be structured, we might have to do it manually!

For this example, we first check the datatypes:


In [30]:
df.dtypes

County        object
ElecYear     float64
Election      object
ElecDate      object
PrecNo       float64
PrecName      object
Party         object
Contest       object
Descr        float64
VoteFor       object
Candidate     object
Posn         float64
Count        float64
PctCnt       float64
Total        float64
PctTot       float64
dtype: object

In [31]:
df.shape

(3526, 16)

A common cause of errors is having Null or 'None' values. This usually means data was missing. 

In [32]:
df[df['PrecNo'].isnull()]

County  ElecYear Election ElecDate  PrecNo PrecName Party Contest  Descr  \
3520    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   
3521    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   
3522    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   
3523    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   
3524    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   
3525    NaN       NaN      NaN      NaN     NaN      NaN   NaN     NaN    NaN   

     VoteFor Candidate  Posn  Count  PctCnt  Total  PctTot  
3520     NaN       NaN   NaN    NaN     NaN    NaN     NaN  
3521     NaN       NaN   NaN    NaN     NaN    NaN     NaN  
3522     NaN       NaN   NaN    NaN     NaN    NaN     NaN  
3523     NaN       NaN   NaN    NaN     NaN    NaN     NaN  
3524     NaN       NaN   NaN    NaN     NaN    NaN     NaN  
3525     NaN       NaN   NaN    NaN     NaN    NaN     NaN

We had some incomplete data! We can get rid of it with the `.notnull()` function.

In [33]:
df = df[df['PrecNo'].notnull()]
df.shape

(3520, 16)

We removed those 6 rows. Important note, this is working in memory! We are not actually changing the table, do to so, you can export your output to a CSV.

### 6. Split-apply-combine (Joins!)

With Pandas we can split the data into groups based on some criteria. We can then apply a function to each group independently, and finally combine the results into a new data structure.

![Joins](http://i.imgur.com/yjNkiwL.png)

#### Group Registered Voter Data and Compare Numbers

In this next section, let's look at registered voter data, and do some grouping, joins, and comparisons.


In [34]:
# Import the file
# Note: It's a big one! You can work with large files in Pandas with ease.
# However, you might need to set the low memory option to False.
df_voters = pd.read_table('data/CENTRE_FVE_20170123.csv', sep=',', low_memory=False)
df_voters

1    2            3            4       5    6    7  \
0       002013865-14  NaN        MILLS        JAMES  ARTHUR  NaN    M   
1       002020582-14  NaN        TERRY      CHARLES       D  NaN    U   
2       002287451-14  NaN      MURRELL      RICHARD       L  NaN    M   
3       002365541-14  NaN        MILLS        JOYCE   ELLEN  NaN    F   
4       002370224-14  NaN       FENTON        CASEY   JAMES  NaN    M   
5       002572852-14  NaN      MURRELL        SUSAN       L  NaN    F   
6       004529096-14  NaN         HAIR       STEVEN       M  NaN    M   
7       005138406-14  NaN       BROOKS        HARRY       C  NaN    M   
8       005163451-14  NaN       BROOKS        DEBRA  LOUISE  NaN    F   
9       005194526-14  NaN    GENSIMORE         GARY   DAVID  NaN    M   
10      005881492-14  NaN       GRAHAM      KATHRYN     LEE  NaN    F   
11      007026230-14  NaN    SCHNEIDER       WALTER       G  NaN    M   
12      007094976-14  NaN       MURRAY      RACHAEL       R  NaN    U   
13      007376473-14  NaN        OCKER      ROSALIE    JEAN  NaN    F   
14      007398108-14  NaN        OCKER         JEAN       E  NaN    F   
15      009621382-14  NaN     MACHIKAS    ALEXANDRA       M  NaN    F   
16      009878128-14  NaN     BRUNGARD         LARA       C  NaN    F   
17      010437417-14  NaN       RANKIN        LUCAS  ROBERT  NaN    M   
18      010464969-14  NaN        SHENK        COLBY    RYAN  NaN    M   
19      010573042-14  NaN     STEPHENS         ERIN       L  NaN    F   
20      010634062-14  NaN       STEVER       SHEILA       M  NaN    F   
21      010849083-14  NaN        SMITH      ANNETTE       L  NaN    F   
22      010849366-14  NaN        MENNA     FRANKLIN       M  NaN    U   
23      010855262-14  NaN       PROVAN        DAVID       A  NaN    U   
24      010857411-14  NaN       WALKER        RALPH       E  NaN    U   
25      010858336-14  NaN      NASTASE    ELIZABETH       E  NaN    U   
26      010858340-14  NaN     BEIGHTOL        VELMA       F  NaN    U   
27      010858347-14  NaN     BONFATTO         MARY       M  NaN    U   
28      010858364-14  NaN        MANEY       DENNIS       C  NaN    U   
29      010858367-14  NaN       RHOADS         DALE       F  NaN    U   
...              ...  ...          ...          ...     ...  ...  ...   
123413  108229919-14  NaN       MISHRA       SONIKA     NaN  NaN    F   
123414  108245924-14  NaN      PECKYNO  CHRISTOPHER       R  NaN  NaN   
123415  108251099-14  NaN     AVEZZANO    KATHERINE       G  NaN  NaN   
123416  108251824-14  NaN      TERRERI       RONALD       C   JR  NaN   
123417  108262483-14  NaN       MADIAS       STEVEN       P  NaN  NaN   
123418  108262546-14  NaN        SACCO    NICOLETTE       D  NaN  NaN   
123419  108263616-14  NaN      SKAVERY    CHANTELLE  NICOLE  NaN    F   
123420  108266993-14  NaN    FROMSTEIN       ELISHA       M  NaN  NaN   
123421  108268621-14  NaN   MCGARRIGAL        EMILY       R  NaN  NaN   
123422  108268872-14  NaN       FELBER          ELI     NaN  NaN  NaN   
123423  108268974-14  NaN      NEWCOMB      MICHAEL       F  III  NaN   
123424  108272412-14  NaN       COLONY       KAYLEY   RENEE  NaN    F   
123425  108273735-14  NaN         MUIR      RACHAEL     NaN  NaN    F   
123426  108281100-14  NaN        BREEN        MAURA       P  NaN    F   
123427  108281392-14  NaN       MENZEL        MEGAN     ANN  NaN    F   
123428  108281516-14  NaN        COLON        ERICA       N  NaN    F   
123429  108283090-14  NaN       SHAHID         AIMA     NaN  NaN    F   
123430  108283218-14  NaN         COPE      ZACHARY     NaN  NaN    M   
123431  108286601-14  NaN       LANDIS         JODY    JEAN  NaN    F   
123432  108286706-14  NaN        ZARRO      ABIGAIL   HAYES  NaN    F   
123433  108288642-14  NaN       BUTLER        JULIA       K  NaN  NaN   
123434  108290220-14  NaN     MCGOVERN        JAMES       J   IV    M   
123435  108290356-14  NaN       GORDON       MARLEY       R  NaN    F   
1234

In [35]:
# Get the row count
df_voters.shape

(123443, 153)

In [36]:
# change column datatype to int for Field 27
df_voters['27'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85,
       86, 87, 88, 89, 90, 91])

Our table is 123443 rows long, by 153 columns. For a data dictionary, see the data folder. Looking at our metadata, our precinct code is held in Field 27.

In [37]:
# Group the edges by their source
grouped = df_voters.groupby('27')
summed = grouped['27'].count()

# get a series
summed.head()

27
1     505
2     846
3     710
4    1016
5    1387
Name: 27, dtype: int64

In [38]:
# Convert our series to a dataframe so we can work with it further
df_summed_extract = summed.to_frame()
# df_summed_extract.index = df_summed_extract.index.map(int)
df_summed_extract.head()

27
27      
1    505
2    846
3    710
4   1016
5   1387

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.rename.html

In [39]:
# lets rename things so they make more sense
df_summed_extract.columns = ['Count']
df_summed_extract.index.names = ['Precinct']
df_summed_extract

Count
Precinct       
1           505
2           846
3           710
4          1016
5          1387
6           892
7           504
8           716
9           580
10          585
11          602
12          529
13          383
14          461
15          265
16          888
17          783
18         3320
19         2426
20         1515
21         1121
22         1150
23         1150
24         4558
25         3644
26         4540
27         4741
28         2420
29         2500
30         1543
...         ...
62          493
63          205
64         2744
65         2381
66         1079
67         2933
68         2209
69          658
70         1052
71         1522
72         1297
73          264
74          123
75          484
76          748
77          364
78         1307
79         1054
80          661
81          536
82          901
83         1124
84         2136
85          531
86         1194
87         1264
88         1392
89         1407
90         1236
91          929

[91 rows x 1 columns]

Next, lets perform a join. We are going to compare the number of registered voters in the voter registration dataset to the same number in the results dataset. Let's join this table we just made from the voter registration dataset to the results dataset, based on precinct number.

How might we do this?

In [40]:
# Query out the necessary rows
df_resultsRegVoters = df.loc[df['Contest'] == 'REGISTERED VOTERS - TOTAL']

# Create our table with headers and new index for joining
df_summed_resultsRV = df_resultsRegVoters[['PrecNo', 'Count']].astype(int).set_index('PrecNo')
df_summed_resultsRV.columns = ['Count']

# View our table
df_summed_resultsRV.head()

Count
PrecNo       
1         507
2         853
3         697
4        1011
5        1378

In [41]:
# run a join!
df_compare = df_summed_extract.join(df_summed_resultsRV, lsuffix='_Extract', rsuffix='_Results')
df_compare

Count_Extract  Count_Results
Precinct                              
1                   505            507
2                   846            853
3                   710            697
4                  1016           1011
5                  1387           1378
6                   892            892
7                   504            504
8                   716            712
9                   580            581
10                  585            592
11                  602            592
12                  529            527
13                  383            384
14                  461            460
15                  265            270
16                  888            883
17                  783            788
18                 3320           3320
19                 2426           2430
20                 1515           1517
21                 1121           1128
22                 1150           1162
23                 1150           1154
24                 4558           4573
25                 3644           3659
26                 4540           4546
27                 4741           4753
28                 2420           2413
29                 2500           2514
30                 1543           1541
...                 ...            ...
62                  493            488
63                  205            205
64                 2744           2728
65                 2381           2389
66                 1079           1078
67                 2933           2946
68                 2209           2206
69                  658            660
70                 1052           1057
71                 1522           1522
72                 1297           1295
73                  264            265
74                  123            125
75                  484            480
76                  748            745
77                  364            364
78                 1307           1290
79                 1054           1045
80                  661            654
81                  536            536
82                  901            903
83                 1124           1122
84                 2136           2122
85                  531            523
86                 1194           1186
87                 1264           1262
88                 1392           1385
89                 1407           1403
90                 1236           1234
91                  929            921

[91 rows x 2 columns]

In [42]:
df_compare['difference'] = np.subtract(df_compare['Count_Extract'], df_compare['Count_Results'])
df_compare

Count_Extract  Count_Results  difference
Precinct                                          
1                   505            507          -2
2                   846            853          -7
3                   710            697          13
4                  1016           1011           5
5                  1387           1378           9
6                   892            892           0
7                   504            504           0
8                   716            712           4
9                   580            581          -1
10                  585            592          -7
11                  602            592          10
12                  529            527           2
13                  383            384          -1
14                  461            460           1
15                  265            270          -5
16                  888            883           5
17                  783            788          -5
18                 3320           3320           0
19                 2426           2430          -4
20                 1515           1517          -2
21                 1121           1128          -7
22                 1150           1162         -12
23                 1150           1154          -4
24                 4558           4573         -15
25                 3644           3659         -15
26                 4540           4546          -6
27                 4741           4753         -12
28                 2420           2413           7
29                 2500           2514         -14
30                 1543           1541           2
...                 ...            ...         ...
62                  493            488           5
63                  205            205           0
64                 2744           2728          16
65                 2381           2389          -8
66                 1079           1078           1
67                 2933           2946         -13
68                 2209           2206           3
69                  658            660          -2
70                 1052           1057          -5
71                 1522           1522           0
72                 1297           1295           2
73                  264            265          -1
74                  123            125          -2
75                  484            480           4
76                  748            745           3
77                  364            364           0
78                 1307           1290          17
79                 1054           1045           9
80                  661            654           7
81                  536            536           0
82                  901            903          -2
83                 1124           1122           2
84                 2136           2122          14
85                  531            523           8
86                 1194           1186           8
87                 1264           1262           2
88                 1392           1385           7
89                 1407           1403           4
90                 1236           1234           2
91                  929            921           8

[91 rows x 3 columns]

The numbers are really close, but not quite the same. Why might this be?